[WIP]Embed feature in demo page

crisner opened the pull request:
Fixes #252  (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

crisner commented:
Screenshot of the feature:
![embedFeature](https://user-images.githubusercontent.com/29401459/70433043-ad250800-1aa6-11ea-84eb-3baf1a2ed4c7.png)

crisner commented:
I have added the embed feature only to the layers.js file as it was specifically for the demo page. Should this be implemented in AllLayers.js as well?

Also, when I was referring to the embed feature implemented in plots2 I noticed that the URL has query strings. Should that be implemented here as well?

crisner commented:
Should the embed button follow the style guide and have a bootstrap styled button or should it be similar to the zoom controls in the map?

crisner commented:
Embed code in use at the bottom of the page --> https://codepen.io/crisren/full/abzdJPV

sagarpreet-chadha commented on src/util/embedControl.js:
```diff
@@ -0,0 +1,50 @@
+
+    generateCode: function() {
+        var params = this.getUrlHashParameters();
+        var code = '<iframe style="border:none;" width="100%" height="900px" src="//publiclab.github.io/leaflet-environmental-layers/example/#' + params.zoom + '/' + params.lat + '/' + params.lng + '/' + params.layers +'"></iframe>';
```
Hey @crisner ,
Instead of this URL, let's use the one i created in PL:
https://publiclab.org/map/

crisner replied:
I tried loading the mentioned URL but I am getting a 404 page.

sagarpreet-chadha commented on src/util/embedControl.js:
```diff
@@ -0,0 +1,50 @@
+        return code;
+    },
+
+    onRemove: function(map) {}
```
Can you verify if on calling remove function, the button gets removed?

crisner replied:
I didn't implement a remove function thinking there wasn't a need for it. I shall get it done.

crisner replied:
The button gets removed using `removeControl(control)` on the map object. It looks like the `onRemove` method works as expected.

sagarpreet-chadha commented on example/layers.js:
```diff
@@ -232,6 +232,8 @@
     var leafletControl = new L.control.layers(baseMaps,overlayMaps);
     leafletControl.addTo(map);
-    
+
```
AWESOME!!!
Let's add this to all HTML demo files :)

sagarpreet-chadha commented on example/index.html:
```diff
@@ -38,6 +38,8 @@
 
   <script src="lib/glify.js"></script>
+
+  <script src="https://kit.fontawesome.com/92fa0312fd.js" crossorigin="anonymous"></script>
```
Can we download this font-awesome file and add that to LEL repo in /images folder so that the button image is there even if we are offline?

sagarpreet-chadha replied:
And then remove this script also. Thanks!

jywarren replied:
We could include it using NPM and then it'd be in `node_modules`?

jywarren commented:
Regarding the styling, i guess it might work best as a native Leaflet styled button, like the full screen buttons on https://publiclab.org/wiki/unearthing-pvd for example. Once you try it out can you share a screenshot? 

Thank you! Looking awesome!

crisner commented:
Thank you for the feedback @sagarpreet-chadha  and @jywarren! I shall make the changes as suggested.

crisner commented:
@jywarren, here's the screenshot of the embed button I have now. What do you think?
![127 0 0 1_5500_example_index html (2)](https://user-images.githubusercontent.com/29401459/70534117-61df2800-1b80-11ea-9aad-5387a7c00108.png)

jywarren commented:
Oh that's beautiful! Great job!

I was thinking that, in the online gh-pages demo, we may want a page that
explicitly says something like "Pan & zoom the map below and select layers
to display, then use the `</>` button to generate an embed code you can use
on your own website." -- this could just be a strip of text at the top, you
know? That way the gh-pages demo is a utility as well as a demo.

Later, we could actually offer a choice of embed types - HTML <iframe> or
plots2 shortcode ([map:______] style) on this example page. What do you
think?



On Tue, Dec 10, 2019 at 8:40 AM Renisha Christie. A <
notifications@github.com> wrote:

> @jywarren <https://github.com/jywarren>, here's the screenshot of the
> embed button I have now. What do you think?
> [image: 127 0 0 1_5500_example_index html (2)]
> <https://user-images.githubusercontent.com/29401459/70534117-61df2800-1b80-11ea-9aad-5387a7c00108.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/leaflet-environmental-layers/pull/296?email_source=notifications&email_token=AAAF6J4EDMYCACKJZCXFNYDQX6L3JA5CNFSM4JYHWOY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGPIGFA#issuecomment-564036372>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J3X7M6PBFRYCOV4SCDQX6L3JANCNFSM4JYHWOYQ>
> .
>

crisner commented:
This is what I have so far:
- Assuming the URL needed to be formatted as:
`http://localhost:3000/map/Unearthing,eonetFiresLayer,odorreport,asian#lat=30.600093873550072&lon=52.03125&zoom=3`
I tried to implement the hash but with the formatted code by modifying the fullHash library. -> https://github.com/crisner/leaflet-fullUrlHash. I have the following format now:
`http://127.0.0.1:5500/example/index.html#lat=43.00&lon=-83.00&zoom=3&layers=BL1`
- Added fontawesome to node modules
- Added native leaflet style to embed icon

Issues I am facing:
- When I click on the prompt buttons, generated more than once after loading the page, for some reason the location hash gets reset. I will be looking into this tomorrow.

jywarren commented:
This sounds great. Could we split out the domain name as a settable option
defaulting to the gh-pages
https://publiclab.github.io/leaflet-environmental-layers/ but overridable
when LEL is initialized with something like options.hostname or
options.root_url ? I guess it's more complicated than that but maybe it's
worth spinning out a new issue to follow-up on this idea. Sorry to throw
more and more into this PR! We can focus on solving a narrower task here
and defer any extra stuff for a separate issue/PR; that's a best practice
anyways!

Thanks!!

On Tue, Dec 10, 2019 at 11:06 AM Renisha Christie. A <
notifications@github.com> wrote:

> This is what I have so far:
>
>    - Assuming the URL needed to be formatted as:
>
>    http://localhost:3000/map/Unearthing,eonetFiresLayer,odorreport,asian#lat=30.600093873550072&lon=52.03125&zoom=3
>    I tried to implement the hash but with the formatted code by modifying
>    the fullHash library. -> https://github.com/crisner/leaflet-fullUrlHash.
>    I have the following format now:
>
>    http://127.0.0.1:5500/example/index.html#lat=43.00&lon=-83.00&zoom=3&layers=BL1
>    - Added fontawesome to node modules
>    - Added native leaflet style to embed icon
>
> Issues I am facing:
>
>    - When I click on the prompt buttons, generated more than once after
>    loading the page, for some reason the location hash gets reset. I will be
>    looking into this tomorrow.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/leaflet-environmental-layers/pull/296?email_source=notifications&email_token=AAAF6J24CUGB2U3PFRLPRKLQX64OZA5CNFSM4JYHWOY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGPYLOY#issuecomment-564102587>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J63UQY4VSPTYKWXQRDQX64OZANCNFSM4JYHWOYQ>
> .
>

crisner commented:
Thanks, @jywarren! I like the idea of spinning it out as a new issue. That way it is easier for me to focus.

jywarren commented:
Super!

On Tue, Dec 10, 2019, 11:25 AM Renisha Christie. A <notifications@github.com>
wrote:

> Thanks, @jywarren <https://github.com/jywarren>! I like the idea of
> spinning it out as a new issue. That way it is easier for me to focus.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/leaflet-environmental-layers/pull/296?email_source=notifications&email_token=AAAF6J3BR57T4QSDAIC5T5DQX67HTA5CNFSM4JYHWOY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGP3CUI#issuecomment-564113745>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J7QCGNFE2NJI3APG4LQX67HTANCNFSM4JYHWOYQ>
> .
>

crisner commented:
- > Issues I am facing:
  > 
  >  When I click on the prompt buttons, generated more than once after loading the page, for some reason the location hash gets reset. I will be looking into this tomorrow.

   Fixed this issue.

- I tried changing this url:
  `http://127.0.0.1:5500/example/index.html#lat=43.00&lon=-83.00&zoom=3&layers=BL1`
  to `http://localhost:3000/example/index.html/BL1#lat=-83.00&lon=-83.00&zoom=3`
  
   I am getting a `cannot GET` error message for `/BL1`.  I am assuming it is because there is no such path on the front end. So, presently the URL is  `http://127.0.0.1:5500/example/index.html#lat=43.00&lon=-83.00&zoom=3&layers=BL1` 

- I have added the embed control to other pages as suggested. This works locally but the actual links point to a 404 page.
![publiclab github io_leaflet-environmental-layers_example_unearthing-pvd html](https://user-images.githubusercontent.com/29401459/70626135-dd58dc00-1c49-11ea-9934-5ddfc44aa2e8.png)

   These files have the permalinks issue [#6889](https://github.com/publiclab/plots2/issues/6889 ) for which we have a PR #293

crisner commented:
Looking forward to your feedback on the changes @sagarpreet-chadha. :) I would also like your feedback on a couple of issues I have mentioned in my previous comment.

crisner commented:
Should I also commit the package-lock.json file?

jywarren commented:
Just quickly, yes please do commit it, as it often includes updates or adjustments. We can pull it out if there's anything too weird in there :-)

Do you know how to test by setting up your own gh-pages branch? This is a good way to check that it'll work the same way without needing to first overwrite the publiclab gh-pages. But you do need to keep a separate branch with gh-pages with your node_modules all added, and commit and push them only to your own gh-pages branch. This gets a little complicated, but it's great for showing off a new version so it can be tested out before being merged!

sagarpreet-chadha commented:
Awesome work so far!!! 
@crisner ...do tell if you need help setting up gh-pages in your forked repo 😄 .

crisner commented:
I have set up gh-pages before. But it has been a while. I'll give it a go and update my progress here.

sagarpreet-chadha commented:
Awesome!!!

On Wed, Dec 11, 2019 at 10:26 PM Renisha Christie. A <
notifications@github.com> wrote:

> I have set up gh-pages before. But it has been a while. I'll give it a go
> and update my progress here.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/leaflet-environmental-layers/pull/296?email_source=notifications&email_token=ADSCRRJFWYC4O2YIBMGY6U3QYELVJA5CNFSM4JYHWOY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGT2XMA#issuecomment-564636592>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSCRRKSI7GJKJJCQIU3GOLQYELVJANCNFSM4JYHWOYQ>
> .
>

crisner commented:
It was harder than I expected to figure things out but I have managed to set up a working gh-pages from a 'development' branch on my forked repo.

#### Screenshot for index.html:
![ghpages_index](https://user-images.githubusercontent.com/29401459/70709144-32f0bf80-1d02-11ea-916a-a284bc78e0b9.png)
Link: https://crisner.github.io/leaflet-environmental-layers/example/index.html#lat=20.55&lon=9.84&zoom=3&layers=BL1,STruth,OdorR

#### Screenshot for oneLinerCodeExample.html:
![ghpages_oneLinerCodeExample](https://user-images.githubusercontent.com/29401459/70709171-456af900-1d02-11ea-9fc7-d26812dcc962.png)
Link: https://crisner.github.io/leaflet-environmental-layers/example/oneLinerCodeExample.html#lat=38.20&lon=-4.04&zoom=3&layers=odorreport,asian,eonetFiresLayer,Unearthing

#### Screenshot for unearthing-pvd.html:
![ghpages_unearthing-pvd](https://user-images.githubusercontent.com/29401459/70709195-574c9c00-1d02-11ea-8e73-be53bfa17c34.png)
Link: https://crisner.github.io/leaflet-environmental-layers/example/unearthing-pvd.html#lat=41.8300&lon=-71.4400&zoom=10&layers=asian,clouds,Unearthing

crisner commented:
I am also getting the following warning from GitHub in my mail,

> The page build completed successfully, but returned the following warning for the `gh-pages` branch:
>
> It looks like you're using GitHub Pages to distribute binary files. We strongly suggest that you use releases to ship projects on GitHub. Releases are GitHub's way of packaging and providing software to your users. You can think of it as a replacement to using downloads to provide software. We found the following file(s) which may be a good candidate for releases: node_modules/phantomjs-prebuilt/lib/phantom/bin/phantomjs.exe.

crisner commented:
The active baselayer appears unselected in the map's layer control which affects the url hash as well.
Screenshot of the issue:
![baselayer-bug](https://user-images.githubusercontent.com/29401459/70794847-9e9b6100-1dc4-11ea-8540-0544b34eca8a.png)

I started another branch from the changes in this PR as the base and created issue #299  and linked PR #300  to it. Please let me know if this is not the right way to go about adding more changes in a separate PR. Thanks!

sagarpreet-chadha commented on example/lib/leaflet-fullUrlHash.js:
```diff
@@ -0,0 +1,198 @@
+	var HAS_HASHCHANGE = (function() {
+		var doc_mode = window.documentMode;
+		return ('onhashchange' in window) &&
+			(doc_mode === undefined || doc_mode > 7);
```
I think we need to add this to a separate folder outside example. Let's make one lib folder at the root level and  change the script tag above.

crisner replied:
Should I move the whole lib folder to the root or just the fullUrlHash.js file?

sagarpreet-chadha commented:
Continued to https://github.com/publiclab/leaflet-environmental-layers/pull/300#pullrequestreview-332204291

sagarpreet-chadha commented:
Yes the whole lib folder, but we should make note that we will have to change script src for glify.js in plots2 then. Thanks!
